### PR TITLE
Tweak Util module

### PR DESCRIPTION
**Pull request by mwichmann**
Move `UtilTests.py` from the top level to the `Util` package directory, for consistency with the other packages with unittests.

Renamed `Util/types.py` -> `Util/sctypes.py`.  `types` is the name of a stdlib module and it's a bad idea to duplicate it, even though in this case it was "legal" since our file was not at the top level.  (Moving `UtilTests.py` actually made this a real problem)

Class `Selector` is no longer an `OrderedDict`, it just inherits from dict as ordering is now preserved and we never used any extra features of `OrderedDict`.

Fix API doc build - was missing a good bit of Util since it was split into a package.

Moved the import-loop warning to the top of `__init__.py` so it will be more visible.

Fiddly linting, doc-stringing, etc.  Super-fiddly: pylint flags `foo`, `bar` and `baz` as prohibited variable/function/method names.  Actually changed these in `UtilTests.py`.

`SCons.Errors` has been a source of import loops because it imports `SCons.Util`. Now Util is split, directly import from the `SCons.Util.sctypes` submodule the two things `Errors` needs - this may reduce the chance of import problems.

## Contributor Checklist:

* [X] I have created a new test or updated the unit tests to cover the new/changed functionality.
* [X] I have updated `CHANGES.txt` (and read the `README.rst`)
* [X] I have updated the appropriate documentation

### CODE REVIEW

**Comment by bdbaddog** [comment]
One concern here is that out in the wild folks are doing
`from SCons.Util.types import is_string`  or similar instead of doing so from SCons.Util.

So minimally this change should explicitly explained in CHANGES and RELEASE.

I'll do some searching of our more notable users git repos and see what's the situation..

**Comment by mwichmann** [comment]
`SCons.Util` only became a package in 4.5, before that, there was no `SCons.Util.types` to import from. Trying to fix my poor choice at the time.

Edit: it seems unlikely that anyone would have picked up importing this in the last six months: it was created such that the symbols continued to be made available just by importing `SCons.Util`, so there would not have been any "not found" event to trigger someone to go looking. Yes, by looking at the source after seeing the release note you can see that there are now distinct module files in the new package, but I'd be surprised if anyone started using for that reasons.

File: SCons/Util/__init__.py [hunk 1]
```diff
@@ -1278,20 +1326,31 @@ def wait_for_process_to_die(pid) -> None:
...
 class DispatchingFormatter(Formatter):
+    """Logging formatter which dispatches to various formatters."""
 
     def __init__(self, formatters, default_formatter) -> None:
         self._formatters = formatters
         self._default_formatter = default_formatter
 
     def format(self, record):
-        formatter = self._formatters.get(record.name, self._default_formatter)
+        # Search from record's logger up to its parents:
```
Comment (bdbaddog) [inline SCons/Util/__init__.py:1336]: How is this different from previous implementation?

Comment (mwichmann) [reply SCons/Util/__init__.py:1336]: additional search. it's an extended version from the same stack overflow article.

**Pull request merged**